Pushdown of LIMIT clause

darold opened the pull request:
Add pushdown of PostgreSQL LIMIT clause translated into FETCH FIRST n ROWS ONLY on Oracle side. There is two restriction to push down the LIMIT clause, an ORDER BY clause must not be present as well as the OFFSET clause that is not supported by Oracle.

darold commented:
This PR might close #463

laurenz commented on sql/oracle_fdw.sql:
```diff
@@ -448,3 +448,18 @@ EXECUTE stmt('{varlena,nonsense}');
+-- With an ORDER BY clause the limit clause must NOT be pushed down
+EXPLAIN (VERBOSE on, COSTS off) SELECT d FROM typetest1 ORDER BY d LIMIT 2;
+SELECT d FROM typetest1 ORDER BY d LIMIT 2;
+-- With an OFFET clause the limit clause must NOT be pushed down
```
Typo: OFFET

laurenz commented on sql/oracle_fdw.sql:
```diff
@@ -8,7 +8,7 @@ SET client_min_messages = WARNING;
 
 -- TWO_TASK or ORACLE_HOME and ORACLE_SID must be set in the server's environment for this to work
-CREATE SERVER oracle FOREIGN DATA WRAPPER oracle_fdw OPTIONS (dbserver '', isolation_level 'read_committed', nchar 'true');
+CREATE SERVER oracle FOREIGN DATA WRAPPER oracle_fdw OPTIONS (dbserver '//192.168.1.26:1521/orcl', isolation_level 'read_committed', nchar 'true');
```
I have left the server empty deliberately - I am using a local connection.
I could agree to 127.0.0.1, but not to 192.168.1.26

laurenz commented on oracle_fdw.c:
```diff
@@ -728,6 +730,69 @@ _PG_init(void)
+			 * we must be sure that there is no ORDER BY clause that
+			 * appears after the LIMIT clause otherwise we will have
+			 * wrong results.
+			 */
```
First, it seems like you are doing it the other way around: you don't push down a `LIMIT` that is *after* an `ORDER BY`. So the comment should be "we must make sure that there is no ORDER BY before the LIMIT", right?

Second, what is the reason for that? If we push down the `ORDER BY`, why shouldn`t we push down the `LIMIT` too?
I must be missing something, but I don't see how that could produce wrong results.
Are you thinking of differences in the collation?

darold replied:
About the ORDER BY restriction this is because the clause might not return the same order from Oracle than from PostgreSQL. This restriction is wrong for numeric column but not for string column where the sort order can be different. I've taken the choice of the restriction but this could be assume and we can push the LIMIT clause in any cases or control this with an option.

About the dbserver value in the regression test, yes sorry I forgot to remove it before pushing. I will fix that as well as the typo.

Concerning `FETCH FIRST n ROWS ONLY` for Oracle version < 12.2 I can change the clause to use ROWNUM just like I'm doing the the revert transformation from Ora2Pg. I agree about cursor and useless of LIMIT but if I recall well oracle_fdw prefetch the first 100 rows, and limit is often used to just return the first row and as you notice this could also help the Oracle planner to choose a better plan.

darold replied:
Searching for a variable to change the oracle_fdw behavior following the Oracle version I see that there is none. Does it make sens for you if I add a new foreign server option **dbversion** to allow the user to set the major version of the Oracle server? It can also be auto-detected at remote connection time but this will add the execution of an extra remote query. Or I can do both, setting the foreign server option will suppress the remote call.

laurenz replied:
There is the `oracleServerVersion` function that I use for `oracle_diag`.
I think it would be OK to call that in `oracleGetSession` immediately after a connection is established, construct a number like PostgreSQL's `server_version_num` from it and store it in the `oracleSession`.
The overhead would be small, and connections are kept open anyway.
Then one can check the server version whenever necessary.

I understand your concern about `ORDER BY`, but then a `LIMIT` without an `ORDER BY` is mostly useless.
[The code](https://github.com/laurenz/oracle_fdw/blob/620495448681ea3314dd1b9660dc013707d2bfb8/oracle_fdw.c#L843) suggests that `ORDER BY` on string columns is not pushed down anyway (for the same reasons), so I think you don't have to worry about that.

darold replied:
Ah right, I've not paid attention to the ORDER BY clause restriction. So that's fine I will apply the LIMIT clause anyway. Yes I have seen the call to oracleServerVersion in the diagnostic, I will reuse it. I will work on it later today.

laurenz reviewed (changes requested):
FIrst, thanks for the patch.

I have two general concerns with this:

1. `FETCH FIRST n ROWS ONLY` was introduced in Oracle 12.2.
   That means that people using an older release of Oracle would get an error message, which is a regression.
   Perhaps I shouldn't care about such an old release any more, but it makes me uncomfortable.

2. Given the way that PostgreSQL and Oracle work, isn't that unnecessary?
   Oracle always executes queries with a cursor, and we fetch rows as we need them.
   Now because of the "Limit" node in PostgreSQL, and because of PostgreSQL's "fetch rows from execution plan nodes as we need them" architecture, we simply stop fetching rows from the Oracle cursor when we reach the limit without any need to explicitly push down the clause.
   Admittedly, pushing down `LIMIT` can give the Oracle optimizer the right clues.
   Also, what I said above is not totally true, since I use Oracle's row prefetching, so I fetch more rows from the cursor than needed.

iliasaz commented:
My example in #463 may have been a bit too simplistic. In reality I saw the described behavior when there was a complex view (not a single table) in the FROM clause. Oracle went for a drastically different plan with FTS (full table scans) as Oracle optimizer's default behavior is to assume the caller wants all rows. Now, if I put "fetch 10 rows only", the plan would change to efficiently fetch a few rows - index scans, Nested loops, and STOP-COUNT optimization and WHERE predicates pushed down into the view.

Considering when to push down and not, I agree there are choices to make. For <12.2 version, one could consider rownum < N condition, or no push-down at all. 
In presence of ORDER BY and >=12.2, it still makes sense to push down both ORDER BY and FETCH N ROWS.

Thanks!

laurenz commented:
@iliasaz Understood. What you are saying makes sense.

darold commented:
Hi Laurenz,

I have changed the way the LIMIT clause is pushed down to the Oracle server. It is now pushed even if there is an ORDER BY clause and the case of Oracle version prior to 10.2 is now covered by the auto-detection of the Oracle server short version (ex: 8.2 or 12.1). I have stored it in the struct oracleSession so that it can be passed to the deparse* functions.

Prior version 10.2 the LIMIT clause is translated into a ROWNUM condition appended to the WHERE clause instead of a FETCH FIRST N ROWS clause at end of the query.
    
I have also added the support for pushdown  of the LIMIT + OFFSET clause in the form of a  ROWNUM condition of the form:
    
            ROWNUM >= offset_value AND ROWNUM <= offset_value+limit_value
    
This is done whatever is the Oracle version provide that there is an OFFSET clause.

Let me know what you think about this implementation also I have not tested directly with an Oracle instance prior to 10.2 but it should works. Note that I have moved the free of the fdwState->session pointer from oracleGetForeignRelSize() to oracleGetForeignPaths() otherwise it is not possible to obtain the Oracle version from the oracleSession struct at deparse time.

While implementing this I have been surprised to see that GROUP BY clauses are not pushed down:
```
EXPLAIN (VERBOSE on, COSTS off) SELECT d FROM typetest1 GROUP BY d;
                                                         QUERY PLAN                                                      
-----------------------------------------------------------------------------------------------------------------------------
 Group
   Output: d
   Group Key: typetest1.d
   ->  Foreign Scan on public.typetest1
         Output: d
         Oracle query: SELECT /*145d52f5de38a843e3af04dc11eeb80c*/ r1."D" FROM "TYPETEST1" r1 ORDER BY r1."D" ASC NULLS LAST
         Oracle plan: SELECT STATEMENT
         Oracle plan:   SORT ORDER BY
         Oracle plan:     TABLE ACCESS FULL TYPETEST1
(9 rows)
```

Is there any reason for that or is is just because it have not be implemented? In the second case I can try to implement it because it can be the cause of very high performances lost. In the example above I understand that we retrieve all the data sorted to apply the group by locally. Let me know, I will open a new issue to add support to GROUP BY pushdown.

iliasaz commented:
Thanks for the quick patch and thoughtful responses, you guys are awesome!

A quick comment on OFFSET. I don't think this is going to work in Oracle - 

> `ROWNUM >= offset_value AND ROWNUM <= offset_value+limit_value`

Here is an example.
this will return 20 rows out of 50
`select level from dual where rownum < 21 connect by level < 51; `

this will not return any rows 
`select level from dual where rownum < 21 and rownum > 9 connect by level < 51; `
because rownum is a pseudo-function that counts the rows in the actual output.

In order to implement OFFSET, we have to wrap it in another SQL. Not sure it this is feasible.
```
select lvl from (
select row_number() over (order by null) rn, level lvl from dual where rownum < 21 connect by level < 51
) 
where rn > 9;
```

laurenz commented:
I am fine with ac2c2f2244fa46fd07cc44eb225d7f73fd6a5907, but I don't like fe1e62d4d3c7214f4ba183616c63ae383b931414 much.

- Minor typo: you used 10.2 instead of 12.2 in some places
- Also minor: you committed your Oracle server again :^)
- I don't like the float for the server version. It seems too cute, and I am worried that rounding errors could be a problem.
  I think this should be one or more integers.

I am also worried about using `ROWNUM`. Adding a `WHERE` condition might change the semantics of the query if it is not a plain `SELECT * FROM atable`. I didn't think this through, but is this save in combination with all possible `WHERE` conditions we might push down? Does it play well with join push-down? Perhaps yes, but that would require testing.

I suggest that you stick with `FETCH FIRST n ROWS ONLY` and just not push down with server versions before 12.2.

I have committed 07cb5ffccc620f113e936cf9f07033a5626d92b6 to retrieve the server version the way I would like it. You can call `oracleServerVersion` like `oracle_diag` does it. I hope that helps, and sorry for adding conflicts with your patches.

darold commented:
I don't think there is any problem to push down these ROWNUM additional WHERE conditions outside if a GROUP BY clause is used.  I can add an exclusion for that but I'm also fine to stick with FETCH FIRST n ROWS ONLY and just not push down with server versions before 12.2. I will do that and also remove the code to support OFFSET, anyway the use of OFFSET might be very rare  and this will keep things more simple to maintain.

I will also fix the typo and rewrite the patch to use your commit 07cb5ff.

Ah and how about the GROUB BY clause, is there is any reason to not pushed down? As said above this can be the cause of lot of performances lost. I can work to add the pushdown of the GROUP BY clause next week.

iliasaz commented:
IMO, OFFSET can be dropped as a rare case.
GROUP BY pushdown certainly adds a lot of value. Also, it's perfectly safe, i.e. it won't impact query semantics.

darold commented:
I have created a new PR #465 base on new oracle_fdw code this one can be closed.

laurenz commented:
> Ah and how about the GROUB BY clause, is there is any reason to not pushed down? As said above this can be the cause of lot of performances lost. I can work to add the pushdown of the GROUP BY clause next week.

The only reason aggregates are not pushed down yet is that it is substantially more complicated. As far as I know, you'd have to write a custom scan node, like for join pushdown.

darold commented:
Ok, I will try to have a look at this next week.
